src/os/bluestore/BlueFS.cc

```diff
@@ -2813,11 +2813,15 @@ int BlueFS::_flush_range(FileWriter *h, uint64_t offset, uint64_t length)
     uint64_t x_len = std::min(p->length - x_off, length);
     bufferlist t;
     t.substr_of(bl, bloff, x_len);
+    h->lock.lock();
```

Comment:
Theoretically this lock doesn't protect p from invalidation. At least I don't see any explicit means for such a protection. I'm not sure this is a real problem for regular operations but it still might be an issue for some (future?) operation scenarios.
Don't know how to fix that easily and not sure whether it needs fixing at all. May be we need to introduce per file R/W lock rather than per-reader/per-writer ones?

Reply:
@ifed01 .  Thanks your suggestions. I'll review the related code.

Reply:
@ifed01 i see the pre-reader/per-writer lock of coarser granularity than the per file R/W lock. and i assume only the writer can invalidate `p`, so practically, it is able to protect the file extent in question. am i right?